zerver/lib/test_runner.py

```diff
@@ -491,9 +500,36 @@ def run_tests(self, test_labels: List[str],
         result = self.run_suite(suite)
         self.teardown_test_environment()
         failed = self.suite_result(suite, result)
-        if not failed:
+        failed_thread = []  # type: List[str]
+        if not failed or self.parallel == 1:
             write_instrumentation_reports(full_suite=full_suite, include_webhooks=include_webhooks)
-        return failed, result.failed_tests
+            return failed, result.failed_tests, failed_thread
+
+        # Below is the logic for passing the list of tests ran in a failed
+        # worker, for use in the --replicate option when running test-backend.
+        args = [
+            (index, subsuite[1])
+            for index, subsuite in enumerate(suite.subsuites)
+        ]  # type: List[Tuple[int, List[str]]]
+
+        # We iterate through the subsuites for the failure. When
+        # we find it, assign the index for use in aggregating the
+        # individual tests.
+        index = None
+        for indexed_tests in args:
```

Comment:
Doing `for (index, subsuite_tests) in args` would be a lot more readable.

Reply:
My [comment](https://github.com/zulip/zulip/pull/12271/files#r295124709) makes `args` go away entirely.

Reply:
I agree that `args` is unnecessary. `args` is actually pretty much straight out of django's source code of `ParallelTestSuite`, which was a direct result of  my effort to understand the structure here.